credentials/jwt/jwt_token_file.go

```diff
@@ -0,0 +1,285 @@
+/*
+ *
+ * Copyright 2025 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package jwt implements gRPC credentials using JWT tokens from files.
+package jwt
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/internal/backoff"
+	"google.golang.org/grpc/status"
+)
+
+// jwtClaims represents the JWT claims structure for extracting expiration time.
+type jwtClaims struct {
+	Exp int64 `json:"exp"`
+}
+
+// jwtTokenFileCallCreds provides JWT token-based PerRPCCredentials that reads
+// tokens from a file.
+// This implementation follows the A97 JWT Call Credentials specification.
+type jwtTokenFileCallCreds struct {
+	tokenFilePath string
+
+	// Cached token data
+	mu               sync.RWMutex
+	cachedToken      string
+	cachedExpiration time.Time // Slightly reduced expiration time compared to the actual exp
+
+	// Error caching with backoff
+	cachedError     error            // Cached error from last failed attempt
+	cachedErrorTime time.Time        // When the error was cached
+	backoffStrategy backoff.Strategy // Backoff strategy when error occurs
+	retryAttempt    int              // Current retry attempt number
+	nextRetryTime   time.Time        // When next retry is allowed
+
+	// Pre-emptive refresh mutex
+	refreshMu sync.Mutex
```

Comment:
The use of two locks complicates things significantly with lock ordering and the possibility of deadlocks. Can we avoid it either by using a single lock or by using other primitives like channels to signal that a pre-emptive refresh is ongoing? Your thoughts please.